### PR TITLE
Blog - Policy Objects: How to configure firewalld for routed networks in libvirt

### PR DESCRIPTION
**Pull request by benblasco**
Guest blog upon suggestion from @erig0.  Originally published at the link below.  I have updated it slightly and can pull it from my blog if it is published here.
https://blog.benblasco.com/routed_networks_libvirt_firewalld/

However @erig0 you mentioned that "simply noting that it's a dual post and linking back to your original would work."

Happy to go with whatever the firewalld team prefers!

### CODE REVIEW

File: blog/_posts/2022-02-04-policy-objects-routing-example.md
Comment (erig0) [inline blog/_posts/2022-02-04-policy-objects-routing-example.md:34]: It's not allowed because firewalld is deny by default.

File: blog/_posts/2022-02-04-policy-objects-routing-example.md
Comment (erig0) [inline blog/_posts/2022-02-04-policy-objects-routing-example.md:36]: This bug is closed. Policy objects ARE the enhancement that allows this setup to work. :)

**Review by erig0 — CHANGES_REQUESTED** [review]
Otherwise, lgtm. Thanks again!

File: blog/_posts/2022-02-04-policy-objects-routing-example.md
Comment (benblasco) [reply blog/_posts/2022-02-04-policy-objects-routing-example.md:34]: A silly question perhaps... But is the behaviour different in a masquerading scenario? Just asking so I can get the language right in the post.

File: blog/_posts/2022-02-04-policy-objects-routing-example.md
Comment (benblasco) [reply blog/_posts/2022-02-04-policy-objects-routing-example.md]: Corrected now.

File: blog/_posts/2022-02-04-policy-objects-routing-example.md
Comment (erig0) [reply blog/_posts/2022-02-04-policy-objects-routing-example.md]: No. Masquerade is the same. It's still forwarded/routed and needs to be allowed.

File: blog/_posts/2022-02-04-policy-objects-routing-example.md
Comment (erig0) [inline blog/_posts/2022-02-04-policy-objects-routing-example.md:34]: Specifically, policy objects were added in v0.9.0.

Comment (erig0) [reply blog/_posts/2022-02-04-policy-objects-routing-example.md:34]: > It turns out that the Firewalld+nftables filter_FORWARD chain prevents non-NATed routing

NATed connections will also be denied by default. The forward filtering occurs _before_ the NAT operations. So the forwarded traffic needs to be allowed in both cases: NAT and non-NAT.

**Review by erig0 — COMMENTED** [review]
A couple new comment, but almost ready.

**Comment by erig0** [comment]
@benblasco, I made quite a few changes. I added some clarity to why it happens and reduced the amount of things "opened" for the VMs. Please take a look.

I'm ready to merge. Just waiting on your approval.

**Review by erig0 — APPROVED** [review]

**Comment by lainestump** [comment]
I just want to point out that the original intent of the libvirt zone *was* to allow all traffic in both directions to/from the VMs; it was added as a parallel to the iptables rules added by libvirt itself for routed networks. Here's an excerpt from https://libvirt.org/firewall.html:

  * type=routed

  Allow inbound, but only to our expected subnet. Allow outbound, but only from our expected
  subnet. Allow traffic between guests. Deny all other inbound. Deny all other outbound.

Discussing this with Eric and reviewing old emails just now, I've learned that at the time the libvirt zone was originally created and tested, it worked due to a bug in the rules created by firewalld (which was a good thing, because at the time it wasn't possible to explicitly allow/deny forwarding via firewalld rules). So if we want to preserve historically desired/documented behavior of routed libvirt virtual networks, this should be made the default setup.

**Comment by erig0** [comment]
> worked due to a bug in the rules created by firewalld (which was a good thing, because at the time it wasn't possible to explicitly allow/deny forwarding via firewalld rules)

The behavior [changed in v1.0.0](https://firewalld.org/2021/06/the-upcoming-1-0-0). See section "Default target is now similar to reject". There is a list of reasons why it was fixed.

> So if we want to preserve historically desired/documented behavior of routed libvirt virtual networks, this should be made the default setup.

Oof. If we want to restore that behavior then a policy that allows the forwarding needs to be added (basically the suggestion in this blog post).

**Comment by benblasco** [comment]
Hi @erig0 I made a couple of minor changes in the interests of tidiness but otherwise it's good to go from my point of view.  

Do we want to add some acknowledgement of the bug @lainestump referred to as part of the blog?  I have a feeling that this bug may ultimately be what brought me here, because this all appeared to work without having to create a policy on my other server which was running Fedora 32

**Comment by erig0** [comment]
I'm holding off on merging this blog as it sounds like this should actually work out of the box. See this Fedora 35 bug: https://bugzilla.redhat.com/show_bug.cgi?id=2055706

**Pull request closed without merging**
